control/injector/Injector.php

```diff
@@ -543,14 +543,25 @@ protected function instantiate($spec, $id=null, $type = null) {
 		}
 		$class = $spec['class'];
 
+		// Check type
+		if (!$type) {
+			$type = isset($spec['type']) ? $spec['type'] : null;
+		}
+
 		// create the object, using any constructor bindings
 		$constructorParams = array();
 		if (isset($spec['constructor']) && is_array($spec['constructor'])) {
 			$constructorParams = $spec['constructor'];
 		}
 
-		$factory = isset($spec['factory']) ? $this->get($spec['factory']) : $this->getObjectCreator();
-		$object = $factory->create($class, $constructorParams);
+		// Catch service references
+		if (strpos($class, '%$') === 0) {
+			$otherID = substr($class, 2);
```

Comment:
At a higher level the "raw" dependency config is stuck in an `array('class' => $spec)` so I've opted to keep it simple here and just check the classname.